// react-native-calendars/src/calendar/day/period/index.js
import _ from 'lodash';
import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {TouchableWithoutFeedback, Text, View} from 'react-native';
import {shouldUpdate} from '../../../component-updater';
import LinearGradient from 'react-native-linear-gradient';
import * as defaultStyle from '../../../style';
import styleConstructor from './style';
import Dot from '../dot';


export default class PeriodDay extends Component {
  
  static displayName = 'IGNORE';

  static propTypes = {
    state: PropTypes.oneOf(['selected', 'disabled', 'today', '']), //TODO: deprecate
    marking: PropTypes.any,
    theme: PropTypes.object,
    onPress: PropTypes.func,
    onLongPress: PropTypes.func,
    date: PropTypes.object,
  };

  constructor(props) {
    super(props);

    this.ole = [];

    this.theme = {...defaultStyle, ...(props.theme || {})};
    this.style = styleConstructor(props.theme);
    
    this.markingStyle = this.getDrawingStyle(props.marking || []);
  }

  onPress = () => {
    this.props.onPress(this.props.date);
  }

  onLongPress = () => {
    this.props.onLongPress(this.props.date);
  }

  shouldComponentUpdate(nextProps) {
    const newMarkingStyle = this.getDrawingStyle(nextProps.marking);
    if (!_.isEqual(this.markingStyle, newMarkingStyle)) {
      this.markingStyle = newMarkingStyle;
      return true;
    }

    return shouldUpdate(this.props, nextProps, ['children', 'state', 'marking', 'onPress', 'onLongPress', 'date']);
  }

  getDrawingStyle(marking) {
    const defaultStyle = {textStyle: {}, containerStyle: {}};
    
    if (!marking) {
      return defaultStyle;
    }

    if (marking.disabled) {
      defaultStyle.textStyle.color = this.style.disabledText.color;
    } else if (marking.selected) {
      defaultStyle.textStyle.color = this.style.selectedText.color;
    }

    const resultStyle = ([marking]).reduce((prev, next) => {
      
      if (next.quickAction) { //???
        if (next.first || next.last) {
          prev.containerStyle = this.style.firstQuickAction;
          prev.textStyle = this.style.firstQuickActionText;
          if (next.endSelected && next.first && !next.last) {
            prev.rightFillerStyle = '#c1e4fe';
          } else if (next.endSelected && next.last && !next.first) {
            prev.leftFillerStyle = '#c1e4fe';
          }
        } else if (!next.endSelected) {
          prev.containerStyle = this.style.quickAction;
          prev.textStyle = this.style.quickActionText;
        } else if (next.endSelected) {
          prev.leftFillerStyle = '#c1e4fe';
          prev.rightFillerStyle = '#c1e4fe';
        }
        return prev;
      }

      if (next.status === 'NotAvailable') { //???
        prev.textStyle = this.style.naText;
      }
      
      const color = next.color;
      if (next.startingDay) {
        prev.startingDay = {color};
      }
      if (next.endingDay) {
        prev.endingDay = {color};
      }
      if (!next.startingDay && !next.endingDay) {
        prev.day = {color};
      }
      if (next.textColor) {
        prev.textStyle.color = next.textColor;
      }
      if (marking.customTextStyle) {
        defaultStyle.textStyle = marking.customTextStyle;
      }
      if (marking.customContainerStyle) {
        defaultStyle.containerStyle = marking.customContainerStyle;
      }
      return prev;
    }, defaultStyle);
    return resultStyle;
  }

  render() {
    const {state, marking} = this.props;
    const containerStyle = [this.style.base];
    const textStyle = [this.style.text];
    let leftFillerStyle = {};
    let rightFillerStyle = {};
    let gradientLeftFillerStyle = [];
    let gradientRightFillerStyle = [];
    let fillerStyle = {};
    let fillers;

    if (state === 'disabled') {
      textStyle.push(this.style.disabledText);
    } else if (state === 'today') {
      containerStyle.push(this.style.today);
      textStyle.push(this.style.todayText);
    }

    if (marking) {
      containerStyle.push({
        borderRadius: 18,
        height: 36,
        overflow: 'hidden',
      });

      const flags = this.markingStyle;
      if (flags.textStyle) {
        textStyle.push(flags.textStyle);
      }
      if (flags.containerStyle) {
        containerStyle.push(flags.containerStyle);
      }
      if (flags.leftFillerStyle) {
        leftFillerStyle.backgroundColor = flags.leftFillerStyle;
        gradientLeftFillerStyle = ['#fff', '#fff']
        gradientRightFillerStyle = ['#fff', '#fff']
      }
      if (flags.rightFillerStyle) {
        rightFillerStyle.backgroundColor = flags.rightFillerStyle;
        gradientLeftFillerStyle = ['#4a54df', '#15d4d8']
        gradientRightFillerStyle = ['#4a54df', '#15d4d8']
      }

      if (flags.startingDay && !flags.endingDay) {
        gradientLeftFillerStyle = ['#fff', '#fff']
        gradientRightFillerStyle = ['#4a54df', '#15d4d8']
        leftFillerStyle = {
          backgroundColor: this.theme.calendarBackground
        };
        rightFillerStyle = {
          backgroundColor: flags.startingDay.color
        };
        containerStyle.push({
          backgroundColor: flags.startingDay.color
        });
      } else if (flags.endingDay && !flags.startingDay) {
        gradientLeftFillerStyle = ['#4a54df', '#15d4d8']
        gradientRightFillerStyle = ['#4a54df', '#15d4d8']
        rightFillerStyle = {
          backgroundColor: this.theme.calendarBackground,
        };
        leftFillerStyle = {
          backgroundColor: flags.endingDay.color,
          zIndex: 10,
          left: 0,
        };
        fillerStyle = {
          width: 30,
        }
        containerStyle.push({
          backgroundColor: 'white',          
        });
        textStyle.push({
          backgroundColor: flags.endingDay.color,
          borderBottomRightRadius: 15,
          borderTopRightRadius: 15,
          width: 30,
          height: 25,
          zIndex: 1,
          top: -2,
          paddingLeft: 5,
          paddingTop: 1
        })
      } else if (flags.day) {
        leftFillerStyle = {backgroundColor: flags.day.color, marginTop: -5};
        rightFillerStyle = {backgroundColor: flags.day.color, marginTop: -5};
        fillerStyle = {backgroundColor: flags.day.color, marginTop: 5,
          height: 25};
        if(flags.day.color) {
          gradientLeftFillerStyle = ['#4a54df', '#15d4d8']
          gradientRightFillerStyle = ['#4a54df', '#15d4d8']
        } else {
          gradientLeftFillerStyle = ['#fff', '#fff']
          gradientRightFillerStyle = ['#fff', '#fff']
          
        }
      } else if (flags.endingDay && flags.startingDay) {
        rightFillerStyle = {
          backgroundColor: this.theme.calendarBackground
        };
        leftFillerStyle = {
          backgroundColor: this.theme.calendarBackground
        };
        gradientLeftFillerStyle = ['#4a54df', '#15d4d8']
        gradientRightFillerStyle = ['#4a54df', '#15d4d8']
        containerStyle.push({
          backgroundColor: flags.endingDay.color
        });
      }

      fillers = (
        <View style={[this.style.fillers, fillerStyle]}>
          <LinearGradient
            start={{x: 0, y: 0}}
            end={{x: 1, y: 0}}
            colors={gradientLeftFillerStyle}
            style={[this.style.leftFiller, leftFillerStyle]}
            useAngle={true}
            angle={261}
            angleCenter={{x: 0.8, y: 0}}>
          </LinearGradient>
          <LinearGradient
            start={{x: 0, y: 0}}
            end={{x: 1, y: 0}}
            colors={gradientRightFillerStyle}
            style={[this.style.rightFiller, rightFillerStyle]}
            useAngle={true}
            angle={261}
            angleCenter={{x: 0.8, y: 0}}>
          </LinearGradient>
          {/* <View style={[this.style.leftFiller, leftFillerStyle]}/>
          <View style={[this.style.rightFiller, rightFillerStyle]}/> */}
        </View>
      );
    }

    const {marking: {marked, dotColor, disableTouchEvent}, theme, accessibilityLabel, testID} = this.props;

    return (
      <TouchableWithoutFeedback
        testID={testID}
        onPress={this.onPress}
        onLongPress={this.onLongPress}
        disabled={disableTouchEvent}
        accessible
        accessibilityRole={disableTouchEvent ? undefined : 'button'}
        accessibilityLabel={accessibilityLabel}
      >
        <View style={this.style.wrapper}>
          {fillers}
          <View style={containerStyle}>
            <Text allowFontScaling={false} style={textStyle}>
              {this.props.children.toString().length === 1 
                ? String(`0${this.props.children}`) 
                : String(this.props.children)}
            </Text>
            <Dot
              theme={theme}
              color={dotColor}
              marked={marked}
            />
          </View>
        </View>
      </TouchableWithoutFeedback>
    );
  }
}
